commit ee2b733e2b5f2389d3473d1cc26bc2ff3070fb4f
Author: zoff99 <zoff99@users.noreply.github.com>
Date:   Sun Apr 24 20:53:15 2022 +0200

    0016_zoff_tc___add_tox_group_peer_count_and_tox_group_get_peerlist_functions

diff --git a/toxcore/group_chats.c b/toxcore/group_chats.c
index 2665df3e9..b05aff470 100644
--- a/toxcore/group_chats.c
+++ b/toxcore/group_chats.c
@@ -8082,6 +8082,25 @@ uint32_t copy_grouplist(const GC_Session *c, uint32_t *out_list, uint32_t list_s
     return ret;
 }
 
+void copy_peerlist(const GC_Chat *chat, uint32_t *out_list)
+{
+    if (out_list == nullptr) {
+        return;
+    }
+
+    if (chat == nullptr) {
+        return;
+    }
+
+    if (chat->numpeers == 0) {
+        return;
+    }
+
+    for (uint32_t i = 0; i < chat->numpeers; ++i) {
+        out_list[i] = chat->group[i].peer_id;
+    }
+}
+
 GC_Chat *gc_get_group(const GC_Session *c, int group_number)
 {
     if (!group_number_valid(c, group_number)) {
diff --git a/toxcore/group_chats.h b/toxcore/group_chats.h
index cfea98fbd..0ea01afcc 100644
--- a/toxcore/group_chats.h
+++ b/toxcore/group_chats.h
@@ -684,6 +684,9 @@ uint32_t gc_count_groups(const GC_Session *c);
 non_null()
 uint32_t copy_grouplist(const GC_Session *c, uint32_t *out_list, uint32_t list_size);
 
+non_null()
+void copy_peerlist(const GC_Chat *chat, uint32_t *out_list);
+
 /** @brief Returns true if peer_number exists */
 non_null()
 bool gc_peer_number_is_valid(const GC_Chat *chat, int peer_number);
diff --git a/toxcore/tox.c b/toxcore/tox.c
index 6098dc815..0106f46e4 100644
--- a/toxcore/tox.c
+++ b/toxcore/tox.c
@@ -3466,6 +3466,52 @@ bool tox_group_self_get_public_key(const Tox *tox, uint32_t group_number, uint8_
     return true;
 }
 
+size_t tox_group_peer_count(const Tox *tox, uint32_t group_number, Tox_Err_Group_Peer_Query *error)
+{
+    assert(tox != nullptr);
+
+    tox_lock(tox);
+    const GC_Chat *chat = gc_get_group(tox->m->group_handler, group_number);
+
+    if (chat == nullptr) {
+        SET_ERROR_PARAMETER(error, TOX_ERR_GROUP_PEER_QUERY_GROUP_NOT_FOUND);
+        tox_unlock(tox);
+        return -1;
+    }
+
+    const size_t ret = chat->numpeers;
+    tox_unlock(tox);
+
+    SET_ERROR_PARAMETER(error, TOX_ERR_GROUP_PEER_QUERY_OK);
+    return ret;
+}
+
+void tox_group_get_peerlist(const Tox *tox, uint32_t group_number, uint32_t *peerlist, Tox_Err_Group_Peer_Query *error)
+{
+    assert(tox != nullptr);
+
+    tox_lock(tox);
+    const GC_Chat *chat = gc_get_group(tox->m->group_handler, group_number);
+
+    if (chat == nullptr) {
+        SET_ERROR_PARAMETER(error, TOX_ERR_GROUP_PEER_QUERY_GROUP_NOT_FOUND);
+        tox_unlock(tox);
+        return;
+    }
+
+    if (peerlist == nullptr) {
+        SET_ERROR_PARAMETER(error, TOX_ERR_GROUP_PEER_QUERY_GROUP_NOT_FOUND);
+        tox_unlock(tox);
+        return;
+    }
+
+    copy_peerlist(chat, peerlist);
+    tox_unlock(tox);
+
+    SET_ERROR_PARAMETER(error, TOX_ERR_GROUP_PEER_QUERY_OK);
+    return;
+}
+
 size_t tox_group_peer_get_name_size(const Tox *tox, uint32_t group_number, uint32_t peer_id,
                                     Tox_Err_Group_Peer_Query *error)
 {
diff --git a/toxcore/tox.h b/toxcore/tox.h
index fb18f20ca..3a77f4fa3 100644
--- a/toxcore/tox.h
+++ b/toxcore/tox.h
@@ -4068,6 +4068,11 @@ typedef enum Tox_Err_Group_Peer_Query {
 } Tox_Err_Group_Peer_Query;
 
 
+size_t tox_group_peer_count(const Tox *tox, uint32_t group_number, Tox_Err_Group_Peer_Query *error);
+
+void tox_group_get_peerlist(const Tox *tox, uint32_t group_number, uint32_t *peerlist, Tox_Err_Group_Peer_Query *error);
+
+
 /**
  * Return the length of the peer's name. If the group number or ID is invalid, the
  * return value is unspecified.
